### ISSUE TITLE
Zoom on cursor rather than center of screen

### ISSUE THREAD

**Issue by sw-yx** [excalidraw/excalidraw#940]
following this Twitter convo: https://twitter.com/Vjeux/status/1238620733812572160

> Seems like it zooms into the center of the page instead of centering the zoom on the cursor. 

and advice from [Gleb:](https://twitter.com/bahmutov/status/1238576418310762496)

> Under the hood it is appending multiplication by a matrix - so to zoom at the cursor location many people forget to first make cursor 0,0 (translation) then scale then apply inverse of the first translation

**Comment by dwelle** [comment]
Related Related #801

**Comment by j-f1** [comment]
<details><summary>Here’s where the code is:</summary>

macOS Safari trackpad pinch:
https://github.com/excalidraw/excalidraw/blob/668f8ec4a6c875e36f2b46934e5bba35da05fc85/src/components/App.tsx#L650-L664

Chrome/FF trackpad pinch, Cmd/Ctrl+scroll:
https://github.com/excalidraw/excalidraw/blob/668f8ec4a6c875e36f2b46934e5bba35da05fc85/src/components/App.tsx#L2040-L2063

Touch pinch: (this might already be OK) 
https://github.com/excalidraw/excalidraw/blob/668f8ec4a6c875e36f2b46934e5bba35da05fc85/src/components/App.tsx#L971-L983

**Comment by j-f1** [comment]
Duplicate of #801

**Comment by sw-yx** [comment]
thanks @j-f1 - taking a crack at this today

**Comment by sw-yx** [comment]
ok i took a crack at this today but didnt come up with a suuper satisfying solution. my solution is quite jumpy.

## Method 1 - modifying `scaleX` and `scaleY`

(i think i'm not sure) the problem is that i don't really have a stable pointer while doing this move and i'm relying on the pointer too much when actually the user's *intent* is to zoom in on an element or arbitrary point in the canvas, and while we zoom that point does not change. is it really a problem of linear algebra? I don't know yet.

I can see how `handleCanvasPointerMove` does it - it waits for two `gesture.pointers` (i suppose this maps to 2 fingers)  and then stores things in `gesture.lastCenter` for diffing next time around so you get a very stable `deltaX` and `deltaY`, which makes it.

If i were to implement this the same way, i'd try to store data on `gesture.lastCenter` as well - i checked and happily enough `handleCanvasPointerMove` doesn't also fire when `handleWheel` fires despite the pointer being on the canvas during the resize. Do you think this is a good idea? (storing a lastCenter and then gradually moving towards it)

I also struggle with converting pointer position (`cursorX`, `cursorY`) to whatever the coordinate system of the canvas is (`scaleX` , `scaleY`). Some hints there might help. I did find `viewportCoordsToSceneCoords` which is used for `getElementAtPosition`, so maybe that's what I need to do?

## Method 2 - modifying `getZoomOrigin`

maybe instead of doing all that complicated math above, we should "just" shift the [Zoom Origin](https://github.com/excalidraw/excalidraw/blob/5dadbcdcd54e4b658512c85d205cccf7e1e9128f/src/scene/zoom.ts#L1) instead from the canvas center? it feels a little easier but could be a little bit more disruptive to the codebase.


I'm *very* new to canvas manipulations, so pardon if I'm asking newbie questions here since I'm also new to this codebase. I dont have to be the one to contribute this fix, if a solution seems obvious to one of you i'm just as happy learning from a PR 😅 Just want to help.

**Comment by pshihn** [comment]
btw, https://github.com/excalidraw/excalidraw/pull/950 should make the zoom code less convoluted when it lands.  It might affect changes you're making.

**Comment by sw-yx** [comment]
thanks, it does look promising. i guess if i use that approach, i just need to understand how to translate from a known cursor position to a point on the canvas, and then gently bump the zoom translation in the direction of the cursor...

**Comment by vjeux** [comment]
Sooooo... we need to change the way zoom has been implemented. Right now when you zoom, it just changes the zoom level and everywhere we compute coordinates it hardcodes the fact that we’re zooming in the middle of the screen.

Instead, what we want is to change the zoom level but also scrollX and scrollY such that we zoom in on the middle of the object. And remove all the assumption in the computation that we zoom in the middle of the screen (all the screenWidth / 2).

Once this refactoring is done, then this becomes a simple task of tweaking the scroll offset based on the cursor location rather than the middle of the screen.

**Comment by sw-yx** [comment]
ok. is there an issue for this somewhere i can follow? sorry i didnt make much progress on this :( i thought of making a PR with what I had, but the jerkiness didn't feel as polished as the rest of excalidraw. idk

**Comment by Jnforja** [comment]
@dwelle Can I get this assigned to me?

**Comment by ReneCode** [comment]
I think **one** transformation matrix for transforming  viewport <-> scene  coords would help.
That matrix would be no big deal, but finding and refactoring all the hard coded translations is a lot of work.

**Comment by Jnforja** [comment]
> I think **one** transformation matrix for transforming viewport <-> scene coords would help.
> That matrix would be no big deal, but finding and refactoring all the hard coded translations is a lot of work.

Thanks for the idea @ReneCode! I'll keep it in mind.

**Comment by sw-yx** [comment]
huge congrats @Jnforja !

**Comment by Jnforja** [comment]
Thanks a lot @sw-yx 😄
